src/include/buffer.h

```diff
@@ -378,6 +378,88 @@ namespace buffer CEPH_BUFFER_API {
       }
     };
 
+    class appender {
+    protected:
+      list *pls;
+      size_t size;
+      ptr bp;
+      char *pos, *end;
+
+    public:
+      appender(bufferlist *l, size_t s) : pls(l), size(s) {
+	prepare_buffer();
+      }
+      appender(appender&& o)
+	: pls(o.pls), size(o.size),
+	  bp(std::move(o.bp)),
+	  pos(o.pos), end(o.end) {
+      }
+      ~appender() {
+	flush();
+      }
+
+      // no copying
+      appender(const appender& other) = delete;
+      appender& operator=(const appender& other) = delete;
+
+      void prepare_buffer(size_t at_least = 0) {
+	bp = buffer::create(std::max(size, at_least));
+	pos = bp.c_str();
+	end = pos + bp.length();
+      }
+
+      void flush() {
+	if (bp.length()) {
+	  bp.set_length(pos - bp.c_str());
+	  pls->append(bp);
+	  bp = ptr();
+	}
+      }
+    };
+
+    class safe_appender : public appender {
+    public:
+      safe_appender(bufferlist *l, size_t s) : appender(l, s) {}
+      void append(char *p, size_t l) {
+	if (pos + l > end) {
+	  flush();
+	  prepare_buffer(l);
+	}
+	memcpy(pos, p, l);
```

Comment:
I'd use maybe_inline_memcpy for that (https://github.com/ceph/ceph/blob/e9e77c855300cee2072f6980ceb6f0e0cda00bbf/src/include/inline_memory.h#L24)